stan/math/rev/mat/functor/algebra_solver.hpp

```diff
@@ -0,0 +1,246 @@
+#ifndef STAN_MATH_REV_MAT_FUNCTOR_ALGEBRA_SOLVER_HPP
+#define STAN_MATH_REV_MAT_FUNCTOR_ALGEBRA_SOLVER_HPP
+
+#include <stan/math/rev/mat/functor/algebra_system.hpp>
+#include <stan/math/prim/mat/fun/mdivide_left.hpp>
+#include <stan/math/rev/core.hpp>
+#include <stan/math/rev/scal/meta/is_var.hpp>
+#include <stan/math/prim/scal/err/check_finite.hpp>
+#include <stan/math/prim/scal/err/check_consistent_size.hpp>
+#include <stan/math/prim/arr/err/check_nonzero_size.hpp>
+#include <unsupported/Eigen/NonLinearOptimization>
+#include <iostream>
+#include <string>
+#include <vector>
+
+namespace stan {
+  namespace math {
+
+    /**
+     * The vari class for the algebraic solver. We compute the  Jacobian of
+     * the solutions with respect to the parameters using the implicit
+     * function theorem. The call to Jacobian() occurs outside the call to
+     * chain() -- this prevents malloc issues.
+     *
+     * Members:
+     * y_ the vector of parameters
+     * y_size_ the number of parameters
+     * x_size_ the number of unknowns
+     * theta_ the vector of solution
+     * Jx_y_ the Jacobian of the solution with respect to the parameters.
+     */
+    template <typename FS, typename F, typename T, typename FX>
+    struct algebra_solver_vari : public vari {
+      vari** y_;
+      int y_size_;
+      int x_size_;
+      vari** theta_;
+      Eigen::MatrixXd Jx_y_;
```

Comment:
Ok. Here's my attempt. I store `Jx_y_` as an std::vector. In the constructor:
```
MatrixXd Jx_y = - mdivide_left(Jf_x, Jf_y);
        double array[x_size_ * y_size_];
        Map<MatrixXd>(&array[0], x_size_, y_size_) = Jx_y;
        Jx_y_ = std::vector<double>(array, array + sizeof(array)
          / sizeof(array[0]));
```
and in the chain method:
```
for (int j = 0; j < y_size_; j++)
          for (int i = 0; i < x_size_; i++)
            y_[j]->adj_ += theta_[i]->adj_
              * Map<MatrixXd>(&Jx_y_[0], x_size_, y_size_)(i, j);
```

Note the line `double array[x_size_ * y_size_];` causes an error with cpplint:
```
Do not use variable-length arrays.  Use an appropriately named ('k' followed by CamelCase) compile-time constant for the size.
```
What does this error message mean?

Also, why do we not want to store an Eigen matrix in a `vari`?